### PR TITLE
Using more readable and shorted method where possible

### PR DESCRIPTION
**Pull request by jamesgua**
Node::isSingleRef()
  Node::isSingleRefUnevaluated()

Issue: #5648 
Signed-off-by: James Guan jamesgua@ca.ibm.com

### CODE REVIEW

**Comment by genie-omr** [comment]
Can one of the admins verify this patch?

**Comment by genie-omr** [comment]
Can one of the admins verify this patch?

**Comment by genie-omr** [comment]
Can one of the admins verify this patch?

**Comment by genie-omr** [comment]
Can one of the admins verify this patch?

**Comment by genie-omr** [comment]
Can one of the admins verify this patch?

**Comment by genie-omr** [comment]
Can one of the admins verify this patch?

**Comment by genie-omr** [comment]
Can one of the admins verify this patch?

**Comment by genie-omr** [comment]
Can one of the admins verify this patch?

**Comment by genie-omr** [comment]
Can one of the admins verify this patch?

**Comment by genie-omr** [comment]
Can one of the admins verify this patch?

**Comment by genie-omr** [comment]
Can one of the admins verify this patch?

**Comment by jamesgua** [comment]
@genie-omr build all

File: compiler/aarch64/codegen/BinaryEvaluator.cpp [hunk 1]
```diff
@@ -51,11 +51,11 @@ genericBinaryEvaluator(TR::Node *node, TR::InstOpCode::Mnemonic regOp, TR::InstO
    TR::Register *trgReg = NULL;
    int64_t value = 0;
 
-   if(1 == firstChild->getReferenceCount())
+   if(true == firstChild->isSingleRef())
```
Comment (fjeremic) [inline compiler/aarch64/codegen/BinaryEvaluator.cpp:54]: These seem redundant. `isSingleRef` returns a boolean value already, so the `true ==` is unnecessary IMO. Could we remove it throughout this PR?

**Comment by 0xdaryl** [comment]
To be honest, I'm not really in favour of using `isSingleRef()` for a couple of reasons (and frankly, I only learned of its existence for the first time today).

First, its semantics aren't clear.  Does `isSingleRef()` mean that this node only has a single reference throughout its lifetime (which may be a question you would ask in the optimizer), or that there is only a single reference remaining after decrementing the reference count (which is may be a question you would ask in the code generator)?  If the semantics are ambiguous and the name equally so then this will lead to mis-use and bugs.

Second, I personally find the code `node->getReferenceCount() == 1` perfectly readable and understandable, and consistent with other reference count checks throughout the codebase.  Does shortening "Reference" to "Ref" make the code that much more readable?  My opinion is no, and if it does then there are many other variables and APIs that should change throughout the compiler.

The API comment for this function also says "A common query used by the optimizer", but I can find no uses of it in the optimizer in OMR nor any known downstream projects.

**Comment by jamesgua** [comment]
I replace _node->getReferenceCount() == 1_ with [isSingleRef](https://github.com/eclipse/omr/blob/master/compiler/il/OMRNode_inlines.hpp#L140) just because it's doing exact the same check _self()->getReferenceCount() == 1_ but shorter name and less verbose, especially in condition checks (e.g. some if checks multiple conditions). If it's discouraged to use isSingleRef then maybe it should disappear in future to avoid usage.

File: compiler/aarch64/codegen/BinaryEvaluator.cpp (hunk 1 above)
Comment (jamesgua) [reply compiler/aarch64/codegen/BinaryEvaluator.cpp]: updated per your suggestion

**Comment by fjeremic** [comment]
I believe #5648 was opened primarily because of `isSingleRefUnevaluated()`. @0xdaryl should we use this API instead of `node->getReferenceCount() == 1 && node->getRegister() == NULL`?

**Comment by 0xdaryl** [comment]
> `isSingleRefUnevaluated()`

Personally, I'm fine with the readability of the code as-is.

However, if something were to be done I think rather than substituting with a literal textual description of the code it is replacing, the code would be made far more readable if it were replaced with the more meaningful higher-level question that is really being asked at that moment.  For instance, is `isOnlyReference()` or `isLoneReference()` the real question being asked?  I suspect in many cases in the code generator it is.  However, the code generator is known to skip evaluation of a node into a register and simply decrement the reference count on occasion (e.g., for constants encodable in an instruction) in which case `isOnlyReference()` isn't strictly correct. 

Without thinking it through in great detail it may be hard to come up with a single API that matches the intent in all cases.  Furthermore, you should be able to use the same function in different parts of the compiler (e.g., optimizer and code generator) and expect the same answer.

**Comment by 0xdaryl** [comment]
And apologies @jamesgua.  We (the community) should have had the discussion to decide on the right path forward in issue #5648 prior to you submitting any code changes.

**Comment by jamesgua** [comment]
close the PR at the moment till the community has a decision.

**Pull request closed without merging**